File: python/mundane/_lock.py
"""Exclusive file locking via flock(2).

Per spec section 3: the runtime acquires flock(LOCK_EX | LOCK_NB) on the
SQLite file's fd. On failure, fail-fast with exit code 75 / LockedError.

Note: flock(2) and POSIX record locks (used internally by SQLite) live in
separate lock-spaces on Linux, so an flock on the DB file does not interfere
with SQLite's own locking.
"""

import contextlib
import fcntl
import os
from typing import Optional


class FileLock:
    """An exclusive non-blocking flock held for the lifetime of the object."""

    def __init__(self, path: str):
        self.path = path
        self._fd: Optional[int] = None

    def acquire(self) -> None:
        # Open for read+write, creating if missing. SQLite will also open it
        # later; that's fine.
        fd = os.open(self.path, os.O_RDWR | os.O_CREAT, 0o644)
        try:
            fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError:
            os.close(fd)
            raise
        self._fd = fd

    def release(self) -> None:
        if self._fd is not None:
            with contextlib.suppress(OSError):
                fcntl.flock(self._fd, fcntl.LOCK_UN)
            with contextlib.suppress(OSError):
                os.close(self._fd)
            self._fd = None

    def __enter__(self) -> "FileLock":
        self.acquire()
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.release()

File: go/mundane/lock.go
package mundane

import (
	"errors"
	"os"
	"strconv"

	"golang.org/x/sys/unix"
)

// FileLock holds an exclusive non-blocking flock(2) on a file.
type FileLock struct {
	path string
	fd   int
}

// AcquireLock opens the file (creating if missing) and takes LOCK_EX|LOCK_NB.
// Returns *LockedError on contention.
func AcquireLock(path string) (*FileLock, error) {
	fd, err := unix.Open(path, unix.O_RDWR|unix.O_CREAT|unix.O_CLOEXEC, 0o644)
	if err != nil {
		return nil, err
	}
	if err := unix.Flock(fd, unix.LOCK_EX|unix.LOCK_NB); err != nil {
		_ = unix.Close(fd)
		if errors.Is(err, unix.EWOULDBLOCK) {
			return nil, &LockedError{Path: path}
		}
		return nil, err
	}
	return &FileLock{path: path, fd: fd}, nil
}

// AdoptLockFromFD wraps an already-flocked fd inherited from a parent process.
// Used when MUNDANE_LOCK_FD is set: we trust the parent holds the lock; the
// child does not re-flock (which would block).
func AdoptLockFromFD(fd int) *FileLock {
	return &FileLock{path: "", fd: fd}
}

// Release unlocks and closes the lock fd. Adopted locks are unlocked by their
// owning parent on exit; the child should not Release.
func (l *FileLock) Release() {
	if l == nil || l.fd <= 0 {
		return
	}
	_ = unix.Flock(l.fd, unix.LOCK_UN)
	_ = unix.Close(l.fd)
	l.fd = -1
}

// SetCloseOnExec marks fd FD_CLOEXEC so a child process (e.g. a step's CMD)
// does not inherit the lock fd and accidentally keep the flock held after the
// shell that owns it exits.
func SetCloseOnExec(fd int) error {
	_, err := unix.FcntlInt(uintptr(fd), unix.F_SETFD, unix.FD_CLOEXEC)
	return err
}

// LockFDFromEnv parses MUNDANE_LOCK_FD; returns -1 if unset or invalid.
func LockFDFromEnv() int {
	n, err := strconv.Atoi(os.Getenv("MUNDANE_LOCK_FD"))
	if err != nil || n < 0 {
		return -1
	}
	return n
}
